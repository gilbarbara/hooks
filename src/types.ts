import { RefObject } from 'react';

export type PlainObject<T = unknown> = Record<PropertyKey, T>;

export type Primitive = bigint | boolean | null | number | string | symbol | undefined;

export type Target<T extends Element> = RefObject<T> | T | null | string;

export type TimerStatus = 'pending' | 'completed' | 'cancelled';
