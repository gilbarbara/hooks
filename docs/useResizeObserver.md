# useResizeObserver

Detect changes in an Element dimensions using the [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver) API.  
Returns a `ResizeObserverEntry`.

## Usage

```tsx
import React from 'react';
import { useResizeObserver } from '@gilbarbara/hooks';

function Component() {
  const ref = React.useRef<HTMLDivElement>(null);
  const entry = useResizeObserver(ref);

  console.log(entry?.contentRect.width);

  return <div ref={ref}>Some content...</div>;
}
```

## Reference

```typescript
useResizeObserver<T extends Element>(
  target: RefObject<T> | T | null | string,
  debounce: number = 0,
): ResizeObserverEntry
```

> This hook uses a [ResizeObserver](https://developer.mozilla.org/en-US/docs/Web/API/ResizeObserver), so if you want to support
legacy browsers, install the [resize-observer-polyfill](https://www.npmjs.com/package/resize-observer-polyfill) in your app.
