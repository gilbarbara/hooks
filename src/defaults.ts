import type { UseMeasureResult } from './useMeasure';

export const defaultElementDimensions: UseMeasureResult = {
  absoluteHeight: 0,
  absoluteWidth: 0,
  bottom: 0,
  height: 0,
  left: 0,
  right: 0,
  top: 0,
  width: 0,
  x: 0,
  y: 0,
};
