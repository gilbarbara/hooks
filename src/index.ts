export { useClickOutside } from './useClickOutside';
export { useDeepCompareEffect } from './useDeepCompareEffect';
export { useDebounce } from './useDebounce';
export { useEffectOnce } from './useEffectOnce';
export {
  useFetch,
  USE_FETCH_STATUS,
  type UseFetchOptions,
  type UseFetchResult,
  type UseFetchStatus,
} from './useFetch';
export { useIntersectionObserver } from './useIntersectionObserver';
export { useInterval } from './useInterval';
export { useIsFirstMount } from './useIsFirstMount';
export { useIsMounted } from './useIsMounted';
export { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
export { useLatest } from './useLatest';
export { useLifecycles } from './useLifecycles';
export {
  useLocalStorage,
  type UseLocalStorageOptions,
  type UseLocalStorageResult,
} from './useLocalStorage';

export {
  useLocalStorageState,
  type UseLocalStorageStateOptions,
  type UseLocalStorageStateResult,
} from './useLocalStorageState';
export { useLocation, type UseLocationResult } from './useLocation';
export { useMeasure, type UseMeasureResult } from './useMeasure';
export { useMediaQuery } from './useMediaQuery';
export { useMergeRefs } from './useMergeRefs';
export { useMount } from './useMount';
export { usePrevious } from './usePrevious';
export { useRenderCount } from './useRenderCount';
export { useResizeObserver } from './useResizeObserver';
export {
  useResponsive,
  type UseResponsiveOrientation,
  type UseResponsiveResult,
} from './useResponsive';
export { useScript, type UseScriptResult } from './useScript';
export { useSetState } from './useSetState';
export { useSingleton } from './useSingleton';
export { useTimeout, type UseTimeoutStatus, type UseTimeoutResult } from './useTimeout';
export { useThrottle } from './useThrottle';
export { useThrottleValue } from './useThrottleValue';
export { useToggle, type UseToggleResult } from './useToggle';
export { useUnmount } from './useUnmount';
export { useUpdate } from './useUpdate';
export { useUpdateEffect } from './useUpdateEffect';
export {
  useWhyDidYouUpdate,
  type UseWhyDidYouUpdateOptions,
  type UseWhyDidYouUpdateResult,
} from './useWhyDidYouUpdate';
export { useWindowSize, type UseWindowSizeResult } from './useWindowSize';
