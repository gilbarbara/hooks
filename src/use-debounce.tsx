import { useEffect, useState } from 'react';

export default function useDebounce(value: unknown, ms: number): unknown {
  // State and setters for debounced value
  const [debouncedValue, setDebouncedValue] = useState(value);

  useEffect(
    () => {
      // Update debounced value after delay
      const handler = setTimeout(() => {
        setDebouncedValue(value);
      }, ms);

      // Cancel the timeout if value changes (also on delay change or unmount)
      // This is how we prevent debounced value from updating if value is changed ...
      // .. within the delay period. Timeout gets cleared and restarted.
      return () => {
        clearTimeout(handler);
      };
    },
    [value, ms], // Only re-call effect if value or delay changes
  );

  return debouncedValue;
}
