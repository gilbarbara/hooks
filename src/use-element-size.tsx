import { useEffect, useLayoutEffect, useMemo, useState } from 'react';

export type UseElementRect = Omit<DOMRectReadOnly, 'toJSON'>;

const defaultState: UseElementRect = {
  x: 0,
  y: 0,
  width: 0,
  height: 0,
  top: 0,
  left: 0,
  bottom: 0,
  right: 0,
};

export default function useElementSize(selector: string): UseElementRect {
  const [element, setElement] = useState<Element | null>();
  const [rect, setRect] = useState<UseElementRect>(defaultState);

  useEffect(() => {
    setElement(document.querySelector(selector));
  }, [selector]);

  const observer = useMemo(
    () =>
      new window.ResizeObserver(entries => {
        if (entries[0]) {
          const { bottom, height, left, right, top, width, x, y } = entries[0].contentRect;

          setRect({ x, y, width, height, top, left, bottom, right });
        }
      }),
    [],
  );

  useLayoutEffect(() => {
    if (!element) {
      return;
    }

    observer.observe(element);

    // eslint-disable-next-line consistent-return
    return () => {
      observer.disconnect();
    };
  }, [observer, element]);

  return rect;
}
