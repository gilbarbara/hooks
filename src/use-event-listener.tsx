import { useEffect, useRef, MutableRefObject } from 'react';

type Handler = (e: Event) => void;

export default function useEventListener(
  eventName: string,
  handler: Handler,
  element: Window | Document | HTMLElement = window,
) {
  const savedHandler: MutableRefObject<null | Handler> = useRef(null);

  useEffect(() => {
    savedHandler.current = handler;
  }, [handler]);

  useEffect(() => {
    const { current } = savedHandler;
    const eventListener = (event: Event) => {
      /* istanbul ignore else */
      if (current) {
        current!(event);
      }
    };

    element.addEventListener(eventName, eventListener);

    return () => {
      element.removeEventListener(eventName, eventListener);
    };
  }, [element, eventName]);
}
