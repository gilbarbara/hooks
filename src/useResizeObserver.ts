import { useMemo, useRef, useState } from 'react';

import { Target } from './types';
import { useIsomorphicLayoutEffect } from './useIsomorphicLayoutEffect';
import { canUseDOM, getElement } from './utils';

export function useResizeObserver<T extends Element>(target: Target<T>, debounce: number = 0) {
  const [element, setElement] = useState(getElement(target));
  const [value, setValue] = useState<ResizeObserverEntry>();
  const timeoutRef = useRef<number | null>(null);

  const isFirstCall = useRef(true);

  const observer = useMemo(() => {
    if (!canUseDOM()) {
      return {};
    }

    return new window.ResizeObserver(entries => {
      if (debounce && !isFirstCall.current) {
        if (timeoutRef.current) {
          window.clearTimeout(timeoutRef.current);
        }

        timeoutRef.current = window.setTimeout(() => {
          setValue(entries[0]);
        }, debounce);

        return;
      }

      setValue(entries[0]);
      isFirstCall.current = false;
    });
  }, [debounce]);

  useIsomorphicLayoutEffect(() => {
    setElement(getElement(target));
  }, [target]);

  useIsomorphicLayoutEffect(() => {
    if (!canUseDOM() || !(observer instanceof ResizeObserver)) {
      return () => undefined;
    }

    if (!element) {
      return () => undefined;
    }

    observer.observe(element);

    return () => {
      observer.disconnect();
    };
  }, [element, observer]);

  return value;
}
