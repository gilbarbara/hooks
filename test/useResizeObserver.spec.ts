import { MutableRefObject } from 'react';
import { act, renderHook } from '@testing-library/react';
import { mockResizeObserver } from 'jsdom-testing-mocks';

import resizeObserverResponse from './__fixtures__/resizeObserverResponse.json';

import { useResizeObserver } from '../src/useResizeObserver';

const resizeObserver = mockResizeObserver();

jest.useFakeTimers();

describe('useResizeObserver', () => {
  const ref = { current: null } as MutableRefObject<Element | null>;
  const rootElement = document.createElement('div');

  rootElement.id = 'root';
  rootElement.innerHTML = 'Hello World';

  beforeAll(() => {
    document.body.appendChild(rootElement);
  });

  beforeEach(() => {
    jest.clearAllMocks();
  });

  afterAll(() => {
    document.body.removeChild(rootElement);
  });

  it('should initialize with an element', () => {
    const { result } = renderHook(() => useResizeObserver(rootElement));

    resizeObserver.mockElementSize(rootElement, resizeObserverResponse);

    act(() => {
      resizeObserver.resize();
    });

    expect(result.current).toMatchSnapshot();
  });

  it('should initialize with a ref', () => {
    const { result } = renderHook(() => useResizeObserver(ref));

    resizeObserver.mockElementSize(rootElement, resizeObserverResponse);

    act(() => {
      resizeObserver.resize();
    });

    expect(result.current).toMatchSnapshot();
  });

  it('should initialize with a string selector and "debounce"', async () => {
    const { rerender, result } = renderHook(() => useResizeObserver('#root', 100));

    resizeObserver.mockElementSize(rootElement, resizeObserverResponse);

    act(() => {
      resizeObserver.resize();
    });

    expect(result.current).toMatchSnapshot('first');

    rerender();

    act(() => {
      resizeObserver.resize(rootElement);
      jest.advanceTimersByTime(50);
    });

    rerender();

    act(() => {
      resizeObserver.resize(rootElement);
      jest.runOnlyPendingTimers();
    });

    expect(result.current).toMatchSnapshot('second');
  });
});
