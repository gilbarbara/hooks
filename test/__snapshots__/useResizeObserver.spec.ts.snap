// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useResizeObserver should initialize with a ref 1`] = `undefined`;

exports[`useResizeObserver should initialize with a string selector and "debounce": first 1`] = `
{
  "borderBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "contentBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "contentRect": {
    "bottom": 200,
    "height": 200,
    "left": 0,
    "right": 200,
    "top": 0,
    "width": 200,
    "x": 0,
    "y": 0,
  },
  "devicePixelContentBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "target": <div
    id="root"
  >
    Hello World
  </div>,
}
`;

exports[`useResizeObserver should initialize with a string selector and "debounce": second 1`] = `
{
  "borderBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "contentBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "contentRect": {
    "bottom": 200,
    "height": 200,
    "left": 0,
    "right": 200,
    "top": 0,
    "width": 200,
    "x": 0,
    "y": 0,
  },
  "devicePixelContentBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "target": <div
    id="root"
  >
    Hello World
  </div>,
}
`;

exports[`useResizeObserver should initialize with an element 1`] = `
{
  "borderBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "contentBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "contentRect": {
    "bottom": 200,
    "height": 200,
    "left": 0,
    "right": 200,
    "top": 0,
    "width": 200,
    "x": 0,
    "y": 0,
  },
  "devicePixelContentBoxSize": [
    {
      "blockSize": 200,
      "inlineSize": 200,
    },
  ],
  "target": <div
    id="root"
  >
    Hello World
  </div>,
}
`;
